cmd/kube-addon-manager/app/addon-manager.go

```diff
@@ -0,0 +1,349 @@
+	}
+	pwf := appendPruneWhitelistFlags(nil, pw)
+	pwf = appendPruneWhitelistFlags(pwf, strings.Fields(env("KUBECTL_EXTRA_PRUNE_WHITELIST")))
+	klog.Infof("== Generated kubectl prune whitelist flags: %s ==", strings.Join(pwf, " "))
+	return pwf
+}
+
+// appendPruneWhitelistFlags generates kubectl prune-whitelist flags from provided resource list.
+func appendPruneWhitelistFlags(pwf, resources []string) []string {
+	for _, resource := range resources {
+		pwf = append(pwf, "--prune-whitelist", resource)
+	}
+	return pwf
+}
+
+func (m *addonManager) Run() error {
+	m.waitForSystemServiceAccount()
+	if err := m.ensureDefaultAdmissionControlsObjects(); err != nil {
+		if os.IsNotExist(err) {
+			klog.Warningf("admissions control object directory not found: %v", err)
+		} else {
+			klog.Errorf("could not create admissions control objects: %v", err)
+		}
+	}
+	for {
+		start := time.Now()
+
+		if !m.leaderElection || m.isLeader() {
+			m.ensureAddons()
+			m.reconcileAddons()
+		} else {
+			klog.Info("== Another addon-manager is acting as leader. Sleeping... ==")
+		}
+
+		elapsed := time.Since(start)
+		if elapsed < m.checkInterval {
+			time.Sleep(m.checkInterval - elapsed)
+		}
+	}
+}
+
+func (m *addonManager) waitForSystemServiceAccount() {
+	var out, errOut bytes.Buffer
+	for {
+		err := m.kubectl(&out, &errOut, m.kubectlOpts, "get", "-n", systemNamespace, "serviceaccount", "default", "-o", "go-template={{with index .secrets 0}}{{.name}}{{end}}")
+		if errStr := errOut.String(); errStr != "" {
+			klog.Warning(errStr)
+		}
+		if err == nil && out.String() != "" {
+			return
+		}
+		out.Reset()
+		errOut.Reset()
+		time.Sleep(500 * time.Millisecond)
+	}
+}
+
+// Create admission_control objects if defined before any other addon services. If the limits
+// are defined in a namespace other than default, we should still create the limits for the
+// default namespace.
+func (m *addonManager) ensureDefaultAdmissionControlsObjects() error {
+	var out, errOut bytes.Buffer
+	return filepath.Walk(m.admissionControlsDir,
+		func(path string, _ os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			ext := filepath.Ext(path)
+			if ext == ".yaml" || ext == ".json" {
+				for attemptsRemaining := startAddonAttempts; attemptsRemaining > 0; attemptsRemaining-- {
+					err = m.kubectl(&out, &errOut, m.kubectlOpts, "--namespace=default", "apply", "-f", path)
+					filterLog(&out, "", klog.Info)
+					filterLog(&errOut, "", klog.Warning)
+					if err == nil {
+						klog.Infof("== Successfully started %s in namespace default ==", path)
+						return nil
+					}
+					klog.Warningf("== Failed to start %s in namespace default. %d tries remaining. ==", path, attemptsRemaining-1)
+					if attemptsRemaining > 1 {
+						time.Sleep(startAddonRetryDelay)
+					}
+				}
+			}
+			return err
+		})
+}
+
+// IsLeader returns whether this instance of addonManager should act as leader in a multi-master cluster.
+func (m *addonManager) isLeader() bool {
```

Comment:
I understand this is meant to keep the implementation at its current form. But has been wondering how easy would it be to implement a real leader election using https://github.com/kubernetes/client-go/blob/master/tools/leaderelection/leaderelection.go?

Reply:
If it's not too bad I'll go ahead and do it, but want to get the direct port working first before implementing improvements